src/acquisition/covidcast/direction_updater.py

```diff
@@ -144,17 +145,135 @@ def get_direction_impl(x, y):
+  df_all = pd.DataFrame(columns=['id', 'source', 'signal', 'time_type', 'geo_type', 'geo_value', 'time_value',
+                                 'timestamp1', 'value', 'timestamp2', 'direction'],
+                        data=database.get_all_record_values_of_timeseries_with_potentially_stale_direction(
+                          tmp_table_name))
+  df_all.drop(columns=['time_type'], inplace=True)
+  df_all['time_value_datetime'] = pd.to_datetime(df_all.time_value, format="%Y%m%d")
+  df_all.direction = df_all.direction.astype(np.float64)
+
+  # Grouping by time-series key, 'time_type' as only time-series with value 'day' were retrieved.
+  groupby_object = df_all.groupby(['source', 'signal', 'geo_type', 'geo_value'])
+
+  num_series = len(groupby_object)
+  num_rows = len(df_all)
+  msg = 'found %d time-series (%d rows) which may have stale direction'
+  print(msg % (num_series, num_rows))
+
+  # get the historical standard deviation of all signals and resolutions
+  rows = database.get_data_stdev_across_locations(
+    Constants.SIGNAL_STDEV_MAX_DAY)
+  data_stdevs = {}
+  for (source, signal, geo_type, aggregate_stdev) in rows:
+    if source not in data_stdevs:
+      data_stdevs[source] = {}
+    if signal not in data_stdevs[source]:
+      data_stdevs[source][signal] = {}
+    data_stdevs[source][signal][geo_type] = aggregate_stdev
+
+  # Dictionary to store the ids of rows with changed direction value
+  changed_rows = {-1.0: [], 0.0: [], 1.0: [], np.nan: []}
+
+  # looping over time-series
+  for ts_index, ts_key in enumerate(groupby_object.groups):
+    (
+        source,
+        signal,
+        geo_type,
+        geo_value,
+    ) = ts_key
+
+    ts_rows = groupby_object.get_group(ts_key).sort_values('time_value_datetime')
+    ts_rows['offsets'] = (ts_rows.time_value_datetime - ts_rows.time_value_datetime.iloc[0]).dt.days
+
+    # progress reporting for anyone debugging/watching the output
+    be_verbose = ts_index < 100
+    be_verbose = be_verbose or ts_index % 100 == 0
+    be_verbose = be_verbose or ts_index >= num_series - 100
+
+    if be_verbose:
+      msg = '[%d/%d] %s %s %s %s: span=%d--%d len=%d max_seconds_stale=%d'
+      args = (
+          ts_index + 1,
+          num_series,
+          source,
+          signal,
+          geo_type,
+          geo_value,
+          ts_rows.time_value.min(),
+          ts_rows.time_value.max(),
+          len(ts_rows),
+          ts_rows.timestamp1.max() - ts_rows.timestamp2.min()
+      )
+      print(msg % args)
+
+    offsets = ts_rows.offsets.values.astype(np.int64)
+    days = ts_rows.time_value.values.astype(np.int64)
+    values = ts_rows.value.values.astype(np.float64)
+    timestamp1s = ts_rows.timestamp1.values.astype(np.int64)
+    timestamp2s = ts_rows.timestamp2.values.astype(np.int64)
+
+    # create a direction classifier for this signal
+    data_stdev = data_stdevs[source][signal][geo_type]
```

Comment:
```suggestion
    data_stdev = data_stdevs[source][signal][geo_type] if (source in data_stdevs) and (signal in data_stdevs[source]) and (geo_type in data_stdevs[source][signal]) else 0
```
To avoid crashing in case some `(source,signal,geo_type)` tuple doesn't have data on or before `Constants.SIGNAL_STDEV_MAX_DAY`.